src/v/utils/timed_mutex.h

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "seastarx.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/semaphore.hh>
+
+#include <utility>
+
+/*
+ * This is a extension of mutex.h. The primary motivation for this class
+ * is to gather statistics on how ofter the mutex is taken (lock_counter)
+ * and the id of the top lock holder.
+ *
+ * Usage
+ * =====
+ *
+ *    ```
+ *    timed_mutex m;
+ *    return m.with("timed_mutex_example", [] { ... });
+ *    ```
+ * The first argument of `with` and `lock` is const char* lock_holder_id.
+ * timed_mutex doesn't make any copy of the string, just pass the pointer
+ * around and may store it indefinely long so it should have unlimited
+ * lifetime. From the caller perspective it means they can't really pass
+ * in a dynamic or stack-allocated string: it should really only be a string
+ * literal. This is important for performance to avoid allocating on every
+ * lock()
+ */
+
+class timed_mutex;
+
+class locked_token {
+    timed_mutex* _mutex;
+    ss::semaphore_units<> _unit;
+    bool _is_moved{false};
+    const char* _lock_holder_id;
+
+public:
+    locked_token(
+      timed_mutex* mutex,
+      const char* lock_holder_id,
+      ss::semaphore_units<>&& unit) noexcept
+      : _mutex(mutex)
+      , _lock_holder_id(lock_holder_id)
+      , _unit(std::move(unit)) {}
+    locked_token(locked_token&& token) noexcept
+      : _mutex(token._mutex)
+      , _lock_holder_id(token._lock_holder_id)
+      , _unit(std::move(token._unit)) {
+        token._is_moved = true;
+    }
+    locked_token(const locked_token&) = delete;
+    locked_token() = delete;
+
+    ~locked_token() noexcept;
+};
+
+struct lock_trace {
+    std::chrono::microseconds duration{0};
+    const char* lock_holder_id{nullptr};
+};
+
+class timed_mutex {
+public:
+    using underlying_t = ss::basic_semaphore<
+      ss::semaphore_default_exception_factory,
+      ss::timer<>::clock>;
+    using duration = typename underlying_t::duration;
+    using time_point = typename underlying_t::time_point;
+
+    explicit timed_mutex(bool should_tracing)
+      : _sem(1)
+      , _is_tracing(should_tracing) {}
+
+    ss::future<> start_tracing() {
```

Comment:
How is `start_tracing` and `stop_tracing` intended to be used? I imagine tracing being turned on/off across the entire system rather than per semaphore. If the intention is that that the timed_mutex is a tool to be used during development but we don't replace all mutexes with timed mutex then I don't think we need start/stop methods?

Reply:
It's a good idea to use it globally but I don't have a final picture in mind, so first I want to replace mutex with timed_mutex in consensus and see how it's going. @Lazin suggested to have global registry of mutexes / semaphores. I prefer to design it based on the experience we get with consensus. With the consensus I envision conrolling timed_mutes (`start_tracing` and `stop_tracing`) based on the config.